Show full call stack in TorchScript exception even when calls were inlined.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29921 Add a pass that reconstructs scopes for nodes from their inline callstacks.
* **#29911 Show full call stack in TorchScript exception even when calls were inlined.**
* #27922 Add logging to inliner.
* #27921 Add InlinedCallStack class.
* #27920 Make inlineCallTo to take Function instead of Graph as the callee argument.
* #27919 Add a variant of insertGraph that fills values map.

Summary:
This uses newly added InlinedCallStack to print the original call stack
even if the real call stack is shallower because of inlining.
This change also makes torchscript stacktraces look like python ones.

Example:
```
@torch.jit.script
def baz(c, b):
    return c + b

@torch.jit.script
def foo(c, b):
    return baz(c, b)

@torch.jit.script
def bar(c, b):
    return foo(c, b)

bar(torch.rand(10), torch.rand(9))
```

Output before:
```
Traceback (most recent call last):
  File "fail.py", line 25, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter, with the following stack trace:
at fail.py:15:11
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output after:
```
Traceback (most recent call last):
  File "fail.py", line 41, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter.
Traceback (most recent call last):
  File "fail.py", line 33
@torch.jit.script
def bar(c, b):
    return foo(c, b)
           ~~~ <--- HERE
  File "fail.py", line 29, in foo
@torch.jit.script
def foo(c, b):
    return baz(c, b)
           ~~~ <--- HERE
  File "fail.py", line 25, in baz
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output of non-scripted python code:
```
Traceback (most recent call last):
  File "fail.py", line 36, in <module>
    bar(torch.rand(10), torch.rand(9))
  File "fail.py", line 21, in bar
    return foo(c, b)
  File "fail.py", line 18, in foo
    return baz(c, b)
  File "fail.py", line 15, in baz
    return c + b
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
```

Differential Revision: [D18532812](https://our.internmc.facebook.com/intern/diff/D18532812)

Commit:
Show full call stack in TorchScript exception even when calls were inlined.

Summary:
This uses newly added InlinedCallStack to print the original call stack
even if the real call stack is shallower because of inlining.

Example:
```
@torch.jit.script
def baz(c, b):
    return c + b

@torch.jit.script
def foo(c, b):
    return baz(c, b)

@torch.jit.script
def bar(c, b):
    return foo(c, b)

bar(torch.rand(10), torch.rand(9))
```

Output before:
```
Traceback (most recent call last):
  File "fail.py", line 25, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter, with the following stack trace:
at fail.py:15:11
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output after:
```
Traceback (most recent call last):
  File "fail.py", line 25, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter, with the following stack trace:
1. at fail.py:23:11
@torch.jit.script
def bar(c, b):
    return foo(c, b)
           ~~~ <--- HERE
2. at fail.py:19:11
@torch.jit.script
def foo(c, b):
    return baz(c, b)
           ~~~ <--- HERE
3. at fail.py:15:11
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1053,6 +1053,14 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
       if (i > 0) {
         out << "during call ";
       }
+      int idx = 1;
```
Can you format the error trace to more closely match python format. e.g. call it a Traceback have the string "(most recent call last)".  
I think it should also mention the function name being called in addition to the source range.

zdevito commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1053,6 +1053,14 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
       if (i > 0) {
         out << "during call ";
```
"during call" formatting here is going to cause the formatting to be different depending of if things have been inlined. I think we should have a test that runs the model with/without inlining and checks the stack trace appears byte-for-byte the same.

zdevito commented on test/test_jit.py:
```diff
@@ -6724,6 +6724,23 @@ def bar(c, b):
+        with self.assertRaises(RuntimeError) as cm:
+            bar(torch.rand(10), torch.rand(9))
+        FileCheck().check("The above operation failed in interpreter, with the following stack trace").check("bar").check("foo").check("baz").run(str(cm.exception))
+
```
We should have a test that includes an interface call (see `test_interface`) to exercise the combined inlined/not-inlined case.

kostmo commented:
## CircleCI build failures summary
As of commit c436b44:
* **1/1** failures introduced in this PR
* **0/1** recognized as flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=c436b44c0df8db52d05555b61447a26c6f92f280):

| Job | Step | Log excerpt |
| --- | --- | --- |
| pytorch_linux_xenial_cuda9_cudnn7_py3_nogpu_test | Test | `main()\':\n/tmp/torch_extensions/test_compilation_error_formatting/main.cpp:2:23: error:` |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
Update on "Show full call stack in TorchScript exception even when calls were inlined."


Summary:
This uses newly added InlinedCallStack to print the original call stack
even if the real call stack is shallower because of inlining.

Example:
```
@torch.jit.script
def baz(c, b):
    return c + b

@torch.jit.script
def foo(c, b):
    return baz(c, b)

@torch.jit.script
def bar(c, b):
    return foo(c, b)

bar(torch.rand(10), torch.rand(9))
```

Output before:
```
Traceback (most recent call last):
  File "fail.py", line 25, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter, with the following stack trace:
at fail.py:15:11
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output after:
```
Traceback (most recent call last):
  File "fail.py", line 25, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter, with the following stack trace:
1. at fail.py:23:11
@torch.jit.script
def bar(c, b):
    return foo(c, b)
           ~~~ <--- HERE
2. at fail.py:19:11
@torch.jit.script
def foo(c, b):
    return baz(c, b)
           ~~~ <--- HERE
3. at fail.py:15:11
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Differential Revision: [D18532812](https://our.internmc.facebook.com/intern/diff/D18532812)

[ghstack-poisoned]

Commit:
Update on "Show full call stack in TorchScript exception even when calls were inlined."


Summary:
This uses newly added InlinedCallStack to print the original call stack
even if the real call stack is shallower because of inlining.
This change also makes torchscript stacktraces look like python ones.

Example:
```
@torch.jit.script
def baz(c, b):
    return c + b

@torch.jit.script
def foo(c, b):
    return baz(c, b)

@torch.jit.script
def bar(c, b):
    return foo(c, b)

bar(torch.rand(10), torch.rand(9))
```

Output before:
```
Traceback (most recent call last):
  File "fail.py", line 25, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter, with the following stack trace:
at fail.py:15:11
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output after:
```
Traceback (most recent call last):
  File "fail.py", line 41, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter.
Traceback (most recent call last):
  File "fail.py", line 33
    return foo(c, b)
  File "fail.py", line 29, in foo
    return baz(c, b)
  File "fail.py", line 25, in baz
    return c + b
```

Output of non-scripted python code:
```
Traceback (most recent call last):
  File "fail.py", line 36, in <module>
    bar(torch.rand(10), torch.rand(9))
  File "fail.py", line 21, in bar
    return foo(c, b)
  File "fail.py", line 18, in foo
    return baz(c, b)
  File "fail.py", line 15, in baz
    return c + b
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
```

Make torchscript stack traces look like python ones.

Differential Revision: [D18532812](https://our.internmc.facebook.com/intern/diff/D18532812)

[ghstack-poisoned]

ZolotukhinM commented:
@zdevito could you please take another look?
Now the stacktraces look as close as possible to python ones. We cannot make them exactly the same, but it's close (see example in commit message). I also added a test with an interface call.

Commit:
Update on "Show full call stack in TorchScript exception even when calls were inlined."


Summary:
This uses newly added InlinedCallStack to print the original call stack
even if the real call stack is shallower because of inlining.
This change also makes torchscript stacktraces look like python ones.

Example:
```
@torch.jit.script
def baz(c, b):
    return c + b

@torch.jit.script
def foo(c, b):
    return baz(c, b)

@torch.jit.script
def bar(c, b):
    return foo(c, b)

bar(torch.rand(10), torch.rand(9))
```

Output before:
```
Traceback (most recent call last):
  File "fail.py", line 25, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter, with the following stack trace:
at fail.py:15:11
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output after:
```
Traceback (most recent call last):
  File "fail.py", line 41, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter.
Traceback (most recent call last):
  File "fail.py", line 33
    return foo(c, b)
  File "fail.py", line 29, in foo
    return baz(c, b)
  File "fail.py", line 25, in baz
    return c + b
```

Output of non-scripted python code:
```
Traceback (most recent call last):
  File "fail.py", line 36, in <module>
    bar(torch.rand(10), torch.rand(9))
  File "fail.py", line 21, in bar
    return foo(c, b)
  File "fail.py", line 18, in foo
    return baz(c, b)
  File "fail.py", line 15, in baz
    return c + b
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
```

Make torchscript stack traces look like python ones.

Differential Revision: [D18532812](https://our.internmc.facebook.com/intern/diff/D18532812)

[ghstack-poisoned]

Commit:
Update on "Show full call stack in TorchScript exception even when calls were inlined."


Summary:
This uses newly added InlinedCallStack to print the original call stack
even if the real call stack is shallower because of inlining.
This change also makes torchscript stacktraces look like python ones.

Example:
```
@torch.jit.script
def baz(c, b):
    return c + b

@torch.jit.script
def foo(c, b):
    return baz(c, b)

@torch.jit.script
def bar(c, b):
    return foo(c, b)

bar(torch.rand(10), torch.rand(9))
```

Output before:
```
Traceback (most recent call last):
  File "fail.py", line 25, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter, with the following stack trace:
at fail.py:15:11
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output after:
```
Traceback (most recent call last):
  File "fail.py", line 41, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter.
Traceback (most recent call last):
  File "fail.py", line 33
    return foo(c, b)
  File "fail.py", line 29, in foo
    return baz(c, b)
  File "fail.py", line 25, in baz
    return c + b
```

Output of non-scripted python code:
```
Traceback (most recent call last):
  File "fail.py", line 36, in <module>
    bar(torch.rand(10), torch.rand(9))
  File "fail.py", line 21, in bar
    return foo(c, b)
  File "fail.py", line 18, in foo
    return baz(c, b)
  File "fail.py", line 15, in baz
    return c + b
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
```

Make torchscript stack traces look like python ones.

Differential Revision: [D18532812](https://our.internmc.facebook.com/intern/diff/D18532812)

[ghstack-poisoned]

Commit:
Update on "Show full call stack in TorchScript exception even when calls were inlined."


Summary:
This uses newly added InlinedCallStack to print the original call stack
even if the real call stack is shallower because of inlining.
This change also makes torchscript stacktraces look like python ones.

Example:
```
@torch.jit.script
def baz(c, b):
    return c + b

@torch.jit.script
def foo(c, b):
    return baz(c, b)

@torch.jit.script
def bar(c, b):
    return foo(c, b)

bar(torch.rand(10), torch.rand(9))
```

Output before:
```
Traceback (most recent call last):
  File "fail.py", line 25, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter, with the following stack trace:
at fail.py:15:11
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output after:
```
Traceback (most recent call last):
  File "fail.py", line 41, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter.
Traceback (most recent call last):
  File "fail.py", line 33
    return foo(c, b)
  File "fail.py", line 29, in foo
    return baz(c, b)
  File "fail.py", line 25, in baz
    return c + b
```

Output of non-scripted python code:
```
Traceback (most recent call last):
  File "fail.py", line 36, in <module>
    bar(torch.rand(10), torch.rand(9))
  File "fail.py", line 21, in bar
    return foo(c, b)
  File "fail.py", line 18, in foo
    return baz(c, b)
  File "fail.py", line 15, in baz
    return c + b
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
```

Make torchscript stack traces look like python ones.

Differential Revision: [D18532812](https://our.internmc.facebook.com/intern/diff/D18532812)

[ghstack-poisoned]

zdevito reviewed (approved):
Nice! Looks good now.

zdevito commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -1044,23 +1044,37 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
-        out << "during call ";
+      if (node->callstack()) {
+        for (const auto& p : (*node->callstack())->vec()) {
+          p.second.print_with_context(
```
I know that Python doesn't add more context by default, but I think we should do it anyway. For instance, python notebooks will print more of the context in stack traces. My previous comments were mostly about the wording differences for the stuff around the highlighting. I also think we should keep the highlighting on, since it makes it easier to read which part of the line we are talking about.

Commit:
Update on "Show full call stack in TorchScript exception even when calls were inlined."


Summary:
This uses newly added InlinedCallStack to print the original call stack
even if the real call stack is shallower because of inlining.
This change also makes torchscript stacktraces look like python ones.

Example:
```
@torch.jit.script
def baz(c, b):
    return c + b

@torch.jit.script
def foo(c, b):
    return baz(c, b)

@torch.jit.script
def bar(c, b):
    return foo(c, b)

bar(torch.rand(10), torch.rand(9))
```

Output before:
```
Traceback (most recent call last):
  File "fail.py", line 25, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter, with the following stack trace:
at fail.py:15:11
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output after:
```
Traceback (most recent call last):
  File "fail.py", line 41, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter.
Traceback (most recent call last):
  File "fail.py", line 33
@torch.jit.script
def bar(c, b):
    return foo(c, b)
           ~~~ <--- HERE
  File "fail.py", line 29, in foo
@torch.jit.script
def foo(c, b):
    return baz(c, b)
           ~~~ <--- HERE
  File "fail.py", line 25, in baz
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output of non-scripted python code:
```
Traceback (most recent call last):
  File "fail.py", line 36, in <module>
    bar(torch.rand(10), torch.rand(9))
  File "fail.py", line 21, in bar
    return foo(c, b)
  File "fail.py", line 18, in foo
    return baz(c, b)
  File "fail.py", line 15, in baz
    return c + b
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
```

Differential Revision: [D18532812](https://our.internmc.facebook.com/intern/diff/D18532812)

[ghstack-poisoned]

Commit:
Update on "Show full call stack in TorchScript exception even when calls were inlined."


Summary:
This uses newly added InlinedCallStack to print the original call stack
even if the real call stack is shallower because of inlining.
This change also makes torchscript stacktraces look like python ones.

Example:
```
@torch.jit.script
def baz(c, b):
    return c + b

@torch.jit.script
def foo(c, b):
    return baz(c, b)

@torch.jit.script
def bar(c, b):
    return foo(c, b)

bar(torch.rand(10), torch.rand(9))
```

Output before:
```
Traceback (most recent call last):
  File "fail.py", line 25, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter, with the following stack trace:
at fail.py:15:11
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output after:
```
Traceback (most recent call last):
  File "fail.py", line 41, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter.
Traceback (most recent call last):
  File "fail.py", line 33
@torch.jit.script
def bar(c, b):
    return foo(c, b)
           ~~~ <--- HERE
  File "fail.py", line 29, in foo
@torch.jit.script
def foo(c, b):
    return baz(c, b)
           ~~~ <--- HERE
  File "fail.py", line 25, in baz
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output of non-scripted python code:
```
Traceback (most recent call last):
  File "fail.py", line 36, in <module>
    bar(torch.rand(10), torch.rand(9))
  File "fail.py", line 21, in bar
    return foo(c, b)
  File "fail.py", line 18, in foo
    return baz(c, b)
  File "fail.py", line 15, in baz
    return c + b
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
```

Differential Revision: [D18532812](https://our.internmc.facebook.com/intern/diff/D18532812)

[ghstack-poisoned]

Commit:
rebase on "Show full call stack in TorchScript exception even when calls were inlined."


Summary:
This uses newly added InlinedCallStack to print the original call stack
even if the real call stack is shallower because of inlining.
This change also makes torchscript stacktraces look like python ones.

Example:
```
@torch.jit.script
def baz(c, b):
    return c + b

@torch.jit.script
def foo(c, b):
    return baz(c, b)

@torch.jit.script
def bar(c, b):
    return foo(c, b)

bar(torch.rand(10), torch.rand(9))
```

Output before:
```
Traceback (most recent call last):
  File "fail.py", line 25, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter, with the following stack trace:
at fail.py:15:11
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output after:
```
Traceback (most recent call last):
  File "fail.py", line 41, in <module>
    bar(torch.rand(10), torch.rand(9))
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
The above operation failed in interpreter.
Traceback (most recent call last):
  File "fail.py", line 33
@torch.jit.script
def bar(c, b):
    return foo(c, b)
           ~~~ <--- HERE
  File "fail.py", line 29, in foo
@torch.jit.script
def foo(c, b):
    return baz(c, b)
           ~~~ <--- HERE
  File "fail.py", line 25, in baz
@torch.jit.script
def baz(c, b):
    return c + b
           ~~~~~ <--- HERE
```

Output of non-scripted python code:
```
Traceback (most recent call last):
  File "fail.py", line 36, in <module>
    bar(torch.rand(10), torch.rand(9))
  File "fail.py", line 21, in bar
    return foo(c, b)
  File "fail.py", line 18, in foo
    return baz(c, b)
  File "fail.py", line 15, in baz
    return c + b
RuntimeError: The size of tensor a (10) must match the size of tensor b (9) at non-singleton dimension 0
```

Differential Revision: [D18532812](https://our.internmc.facebook.com/intern/diff/D18532812)

[ghstack-poisoned]

ZolotukhinM commented:
It's already committed.
